### ISSUE TITLE
Duplicate entries in `/proc/meminfo` make bazel server crash

### ISSUE THREAD

**Issue by DarkDimius** [bazelbuild/bazel#11105]
### Description of the problem / feature request:

running `bazel build //:hello_world` leads to

```
$ bazel build //:hello_world
Starting local Bazel server and connecting to it...

Server terminated abruptly (error code: 14, error message: 'Socket closed', log file: '/home/dark/.cache/bazel/_bazel_dark/37eeb5c8b6e4daaae7f4d5edfdef7735/server/jvm.out')
```

jvm.out:
```
java.lang.IllegalArgumentException: Multiple entries with same key: Writeback=0 and Writeback=0
        at com.google.common.collect.ImmutableMap.conflictException(ImmutableMap.java:215)
        at com.google.common.collect.ImmutableMap.checkNoConflict(ImmutableMap.java:209)
        at com.google.common.collect.RegularImmutableMap.checkNoConflictInKeyBucket(RegularImmutableMap.java:147)
        at com.google.common.collect.RegularImmutableMap.fromEntryArray(RegularImmutableMap.java:110)
        at com.google.common.collect.ImmutableMap$Builder.build(ImmutableMap.java:393)
        at com.google.devtools.build.lib.unix.ProcMeminfoParser.<init>(ProcMeminfoParser.java:62)
        at com.google.devtools.build.lib.actions.LocalHostResourceManagerLinux.getMemoryInMbHelper(LocalHostResourceManagerLinux.java:86)
        at com.google.devtools.build.lib.actions.LocalHostResourceManagerLinux.getMemoryInMb(LocalHostResourceManagerLinux.java:40)
        at com.google.devtools.build.lib.actions.LocalHostResourceManagerLinux.getLocalHostResources(LocalHostResourceManagerLinux.java:46)
        at com.google.devtools.build.lib.actions.LocalHostCapacity.getNewLocalHostCapacity(LocalHostCapacity.java:46)
        at com.google.devtools.build.lib.actions.LocalHostCapacity.getLocalHostCapacity(LocalHostCapacity.java:34)
        at com.google.devtools.build.lib.runtime.LoadingPhaseThreadsOption$LoadingPhaseThreadCountConverter.lambda$new$0(LoadingPhaseThreadsOption.java:54)
        at com.google.devtools.build.lib.util.ResourceConverter.applyOperator(ResourceConverter.java:136)
        at com.google.devtools.build.lib.util.ResourceConverter.convert(ResourceConverter.java:122)
        at com.google.devtools.build.lib.util.ResourceConverter.convert(ResourceConverter.java:41)
        at com.google.devtools.common.options.OptionDefinition.getDefaultValue(OptionDefinition.java:270)
        at com.google.devtools.common.options.OptionValueDescription$DefaultOptionValueDescription.getValue(OptionValueDescription.java:123)
        at com.google.devtools.common.options.OptionsParserImpl.parse(OptionsParserImpl.java:378)
        at com.google.devtools.common.options.OptionsParserImpl.parse(OptionsParserImpl.java:325)
        at com.google.devtools.common.options.OptionsParser.parseWithSourceFunction(OptionsParser.java:642)
        at com.google.devtools.build.lib.runtime.BlazeOptionHandler.parseArgsAndConfigs(BlazeOptionHandler.java:207)
        at com.google.devtools.build.lib.runtime.BlazeOptionHandler.parseOptions(BlazeOptionHandler.java:274)
        at com.google.devtools.build.lib.runtime.BlazeCommandDispatcher.execExclusively(BlazeCommandDispatcher.java:275)
        at com.google.devtools.build.lib.runtime.BlazeCommandDispatcher.exec(BlazeCommandDispatcher.java:208)
        at com.google.devtools.build.lib.server.GrpcServerImpl.executeCommand(GrpcServerImpl.java:603)
        at com.google.devtools.build.lib.server.GrpcServerImpl.lambda$run$2(GrpcServerImpl.java:659)
        at io.grpc.Context$1.run(Context.java:595)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.base/java.lang.Thread.run(Unknown Source)
ERROR: bazel crash in async thread:
java.lang.IllegalArgumentException: Multiple entries with same key: Writeback=0 and Writeback=0
        at com.google.common.collect.ImmutableMap.conflictException(ImmutableMap.java:215)
        at com.google.common.collect.ImmutableMap.checkNoConflict(ImmutableMap.java:209)
        at com.google.common.collect.RegularImmutableMap.checkNoConflictInKeyBucket(RegularImmutableMap.java:147)
        at com.google.common.collect.RegularImmutableMap.fromEntryArray(RegularImmutableMap.java:110)
        at com.google.common.collect.ImmutableMap$Builder.build(ImmutableMap.java:393)
        at com.google.devtools.build.lib.unix.ProcMeminfoParser.<init>(ProcMeminfoParser.java:62)
        at com.google.devtools.build.lib.actions.LocalHostResourceManagerLinux.getMemoryInMbHelper(LocalHostResourceManagerLinux.java:86)
        at com.google.devtools.build.lib.actions.LocalHostResourceManagerLinux.getMemoryInMb(LocalHostResourceManagerLinux.java:40)
        at com.google.devtools.build.lib.actions.LocalHostResourceManagerLinux.getLocalHostResources(LocalHostResourceManagerLinux.java:46)
        at com.google.devtools.build.lib.actions.LocalHostCapacity.getNewLocalHostCapacity(LocalHostCapacity.java:46)
        at com.google.devtools.build.lib.actions.LocalHostCapacity.getLocalHostCapacity(LocalHostCapacity.java:34)
        at com.google.devtools.build.lib.runtime.LoadingPhaseThreadsOption$LoadingPhaseThreadCountConverter.lambda$new$0(LoadingPhaseThreadsOption.java:54)
        at com.google.devtools.build.lib.util.ResourceConverter.applyOperator(ResourceConverter.java:136)
        at com.google.devtools.build.lib.util.ResourceConverter.convert(ResourceConverter.java:122)
        at com.google.devtools.build.lib.util.ResourceConverter.convert(ResourceConverter.java:41)
        at com.google.devtools.common.options.OptionDefinition.getDefaultValue(OptionDefinition.java:270)
        at com.google.devtools.common.options.OptionValueDescription$DefaultOptionValueDescription.getValue(OptionValueDescription.java:123)
        at com.google.devtools.common.options.OptionsParserImpl.parse(OptionsParserImpl.java:378)
        at com.google.devtools.common.options.OptionsParserImpl.parse(OptionsParserImpl.java:325)
        at com.google.devtools.common.options.OptionsParser.parseWithSourceFunction(OptionsParser.java:642)
        at com.google.devtools.build.lib.runtime.BlazeOptionHandler.parseArgsAndConfigs(BlazeOptionHandler.java:207)
        at com.google.devtools.build.lib.runtime.BlazeOptionHandler.parseOptions(BlazeOptionHandler.java:274)
        at com.google.devtools.build.lib.runtime.BlazeCommandDispatcher.execExclusively(BlazeCommandDispatcher.java:275)
        at com.google.devtools.build.lib.runtime.BlazeCommandDispatcher.exec(BlazeCommandDispatcher.java:208)
        at com.google.devtools.build.lib.server.GrpcServerImpl.executeCommand(GrpcServerImpl.java:603)
        at com.google.devtools.build.lib.server.GrpcServerImpl.lambda$run$2(GrpcServerImpl.java:659)
        at io.grpc.Context$1.run(Context.java:595)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.base/java.lang.Thread.run(Unknown Source)
```

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

> Replace this line with your answer.

### What operating system are you running Bazel on?

```
$ uname -a
Linux buildbox 5.3.13-2-pve #1 SMP PVE 5.3.13-2 (Fri, 24 Jan 2020 09:49:36 +0100) x86_64 x86_64 x86_64 GNU/Linux
```

/proc/meminfo

```
MemTotal:       100663296 kB
MemFree:        93340676 kB
MemAvailable:   98430428 kB
Buffers:               0 kB
Cached:          5089752 kB
SwapCached:            0 kB
Active:          4217336 kB
Inactive:        1956280 kB
Active(anon):    1565476 kB
Inactive(anon):   745800 kB
Active(file):    2651860 kB
Inactive(file):  1210480 kB
Unevictable:           0 kB
Mlocked:            5336 kB
SwapTotal:             0 kB
SwapFree:              0 kB
Dirty:                 0 kB
Writeback:             0 kB
AnonPages:        733396 kB
Mapped:            43032 kB
Shmem:           1577880 kB
KReclaimable:    1557508 kB
Slab:               0 kB
SReclaimable:          0 kB
SUnreclaim:            0 kB
KernelStack:       41392 kB
PageTables:       103680 kB
NFS_Unstable:          0 kB
Bounce:                0 kB
Writeback:             0 kB
CommitLimit:    65929152 kB
Committed_AS:   19702672 kB
VmallocTotal:   34359738367 kB
VmallocUsed:     3950392 kB
VmallocChunk:          0 kB
Percpu:           564736 kB
HardwareCorrupted:     0 kB
AnonHugePages:         0 kB
ShmemHugePages:        0 kB
ShmemPmdMapped:        0 kB
CmaTotal:              0 kB
CmaFree:               0 kB
HugePages_Total:       0
HugePages_Free:        0
HugePages_Rsvd:        0
HugePages_Surp:        0
Hugepagesize:       2048 kB
Hugetlb:               0 kB
DirectMap4k:    94660008 kB
DirectMap2M:    37351424 kB
DirectMap1G:     2097152 kB
```

### What's the output of `bazel info release`?
```
$ bazel info release
Starting local Bazel server and connecting to it...

Server terminated abruptly (error code: 14, error message: 'Socket closed', log file: '/home/dark/.cache/bazel/_bazel_dark/37eeb5c8b6e4daaae7f4d5edfdef7735/server/jvm.out')
```

### If `bazel info release` returns "development version" or "(@non-git)", tell us how you built Bazel.

this is official release of 2.2.0

```
$ bazel version
Starting local Bazel server and connecting to it...
Build label: 2.2.0
Build target: bazel-out/k8-opt/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Tue Mar 3 09:26:12 2020 (1583227572)
Build timestamp: 1583227572
Build timestamp as int: 1583227572
```

### What's the output of `git remote get-url origin ; git rev-parse master ; git rev-parse HEAD` ?
This isn't a git repo.

###  Have you found anything relevant by searching the web?
No, but I've found tests
https://github.com/bazelbuild/bazel/blob/1f684e1b87cd8881a0a4b33e86ba66743e32d674/src/test/java/com/google/devtools/build/lib/actions/LocalHostResourceManagerLinuxTest.java

that seem to not cover this case

### Any other information, logs, or outputs that you want to share?
See above

**Comment by aiuto** [comment]
There simply is not enough to triage this.
How did you build the container that causes this?
Is this actually an error in container setup, or a normal use case?

**Comment by Links2004** [comment]
Hi,
looks like I am running in to the same problem.

I have a proxmox server and created a LXC Debain 10 container in it.



```
[lxc]~/bazel_examples/examples/cpp-tutorial/stage1 (master) $ bazel build --verbose_failures //main:hello-world 
Starting local Bazel server and connecting to it...

Server terminated abruptly (error code: 14, error message: 'Socket closed', log file: '/home/sfos/.cache/bazel/_bazel_sfos/3f321be1cafa1793bc6eb1acb8b9c579/server/jvm.out')

[lxc]~/bazel_examples/examples/cpp-tutorial/stage1 (master) $ cat /home/sfos/.cache/bazel/_bazel_sfos/3f321be1cafa1793bc6eb1acb8b9c579/server/jvm.out
java.lang.IllegalArgumentException: Multiple entries with same key: Writeback=1056 and Writeback=1056
	at com.google.common.collect.ImmutableMap.conflictException(ImmutableMap.java:215)
	at com.google.common.collect.ImmutableMap.checkNoConflict(ImmutableMap.java:209)
	at com.google.common.collect.RegularImmutableMap.checkNoConflictInKeyBucket(RegularImmutableMap.java:147)
	at com.google.common.collect.RegularImmutableMap.fromEntryArray(RegularImmutableMap.java:110)
	at com.google.common.collect.ImmutableMap$Builder.build(ImmutableMap.java:393)
	at com.google.devtools.build.lib.unix.ProcMeminfoParser.<init>(ProcMeminfoParser.java:62)
	at com.google.devtools.build.lib.actions.LocalHostResourceManagerLinux.getMemoryInMbHelper(LocalHostResourceManagerLinux.java:86)
	at com.google.devtools.build.lib.actions.LocalHostResourceManagerLinux.getMemoryInMb(LocalHostResourceManagerLinux.java:40)
	at com.google.devtools.build.lib.actions.LocalHostResourceManagerLinux.getLocalHostResources(LocalHostResourceManagerLinux.java:46)
	at com.google.devtools.build.lib.actions.LocalHostCapacity.getNewLocalHostCapacity(LocalHostCapacity.java:46)
	at com.google.devtools.build.lib.actions.LocalHostCapacity.getLocalHostCapacity(LocalHostCapacity.java:34)
	at com.google.devtools.build.lib.runtime.LoadingPhaseThreadsOption$LoadingPhaseThreadCountConverter.lambda$new$0(LoadingPhaseThreadsOption.java:54)
	at com.google.devtools.build.lib.util.ResourceConverter.applyOperator(ResourceConverter.java:136)
	at com.google.devtools.build.lib.util.ResourceConverter.convert(ResourceConverter.java:122)
	at com.google.devtools.build.lib.util.ResourceConverter.convert(ResourceConverter.java:41)
	at com.google.devtools.common.options.OptionDefinition.getDefaultValue(OptionDefinition.java:271)
	at com.google.devtools.common.options.OptionValueDescription$DefaultOptionValueDescription.getValue(OptionValueDescription.java:123)
	at com.google.devtools.common.options.OptionsParserImpl.parse(OptionsParserImpl.java:378)
	at com.google.devtools.common.options.OptionsParserImpl.parse(OptionsParserImpl.java:325)
	at com.google.devtools.common.options.OptionsParser.parseWithSourceFunction(OptionsParser.java:653)
	at com.google.devtools.build.lib.runtime.BlazeOptionHandler.parseArgsAndConfigs(BlazeOptionHandler.java:200)
	at com.google.devtools.build.lib.runtime.BlazeOptionHandler.parseOptions(BlazeOptionHandler.java:267)
	at com.google.devtools.build.lib.runtime.BlazeCommandDispatcher.execExclusively(BlazeCommandDispatcher.java:275)
	at com.google.devtools.build.lib.runtime.BlazeCommandDispatcher.exec(BlazeCommandDispatcher.java:208)
	at com.google.devtools.build.lib.server.GrpcServerImpl.executeCommand(GrpcServerImpl.java:603)
	at com.google.devtools.build.lib.server.GrpcServerImpl.lambda$run$2(GrpcServerImpl.java:663)
	at io.grpc.Context$1.run(Context.java:595)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
ERROR: bazel crash in async thread:
java.lang.IllegalArgumentException: Multiple entries with same key: Writeback=1056 and Writeback=1056
	at com.google.common.collect.ImmutableMap.conflictException(ImmutableMap.java:215)
	at com.google.common.collect.ImmutableMap.checkNoConflict(ImmutableMap.java:209)
	at com.google.common.collect.RegularImmutableMap.checkNoConflictInKeyBucket(RegularImmutableMap.java:147)
	at com.google.common.collect.RegularImmutableMap.fromEntryArray(RegularImmutableMap.java:110)
	at com.google.common.collect.ImmutableMap$Builder.build(ImmutableMap.java:393)
	at com.google.devtools.build.lib.unix.ProcMeminfoParser.<init>(ProcMeminfoParser.java:62)
	at com.google.devtools.build.lib.actions.LocalHostResourceManagerLinux.getMemoryInMbHelper(LocalHostResourceManagerLinux.java:86)
	at com.google.devtools.build.lib.actions.LocalHostResourceManagerLinux.getMemoryInMb(LocalHostResourceManagerLinux.java:40)
	at com.google.devtools.build.lib.actions.LocalHostResourceManagerLinux.getLocalHostResources(LocalHostResourceManagerLinux.java:46)
	at com.google.devtools.build.lib.actions.LocalHostCapacity.getNewLocalHostCapacity(LocalHostCapacity.java:46)
	at com.google.devtools.build.lib.actions.LocalHostCapacity.getLocalHostCapacity(LocalHostCapacity.java:34)
	at com.google.devtools.build.lib.runtime.LoadingPhaseThreadsOption$LoadingPhaseThreadCountConverter.lambda$new$0(LoadingPhaseThreadsOption.java:54)
	at com.google.devtools.build.lib.util.ResourceConverter.applyOperator(ResourceConverter.java:136)
	at com.google.devtools.build.lib.util.ResourceConverter.convert(ResourceConverter.java:122)
	at com.google.devtools.build.lib.util.ResourceConverter.convert(ResourceConverter.java:41)
	at com.google.devtools.common.options.OptionDefinition.getDefaultValue(OptionDefinition.java:271)
	at com.google.devtools.common.options.OptionValueDescription$DefaultOptionValueDescription.getValue(OptionValueDescription.java:123)
	at com.google.devtools.common.options.OptionsParserImpl.parse(OptionsParserImpl.java:378)
	at com.google.devtools.common.options.OptionsParserImpl.parse(OptionsParserImpl.java:325)
	at com.google.devtools.common.options.OptionsParser.parseWithSourceFunction(OptionsParser.java:653)
	at com.google.devtools.build.lib.runtime.BlazeOptionHandler.parseArgsAndConfigs(BlazeOptionHandler.java:200)
	at com.google.devtools.build.lib.runtime.BlazeOptionHandler.parseOptions(BlazeOptionHandler.java:267)
	at com.google.devtools.build.lib.runtime.BlazeCommandDispatcher.execExclusively(BlazeCommandDispatcher.java:275)
	at com.google.devtools.build.lib.runtime.BlazeCommandDispatcher.exec(BlazeCommandDispatcher.java:208)
	at com.google.devtools.build.lib.server.GrpcServerImpl.executeCommand(GrpcServerImpl.java:603)
	at com.google.devtools.build.lib.server.GrpcServerImpl.lambda$run$2(GrpcServerImpl.java:663)
	at io.grpc.Context$1.run(Context.java:595)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
```

```
$ bazel version
Starting local Bazel server and connecting to it...
Build label: 3.1.0
Build target: bazel-out/k8-opt/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Wed Apr 22 10:32:27 2020 (1587551547)
Build timestamp: 1587551547
Build timestamp as int: 1587551547
```

**Comment by ulfjack** [comment]
This should be easy to fix. Do you want to send a patch?

**Comment by Links2004** [comment]
from reading the code the fix is most likely:

```diff
diff --git a/src/main/java/com/google/devtools/build/lib/unix/ProcMeminfoParser.java b/src/main/java/com/google/devtools/build/lib/unix/ProcMeminfoParser.java
index dcfe8086f8..c3c15367c8 100644
--- a/src/main/java/com/google/devtools/build/lib/unix/ProcMeminfoParser.java
+++ b/src/main/java/com/google/devtools/build/lib/unix/ProcMeminfoParser.java
@@ -54,7 +54,9 @@ public class ProcMeminfoParser {
       String valString = line.substring(colon + 1);
       try {
         long val =  Long.parseLong(CharMatcher.inRange('0', '9').retainFrom(valString));
-        builder.put(keyword, val);
+        if (!builder.containsKey(keyword)) {
+          builder.put(keyword, val);
+        }
       } catch (NumberFormatException e) {
         // Ignore: we'll fail later if somebody tries to capture this value.
       }

```

but need to build it for testing, 
which is not possible on this machine do to the bug.

**Comment by tetromino** [comment]
Should be fixed by ae03671e73e0a1933f0514019e33e357b1fb6050 unless I'm mistaken.
